File: backend/wall_fixtures.py
"""Reviewed wall lamps use a wall anchor in the shared centered room."""
from backend.design import require


def is_wall_fixture(product):
    return product.get('placement', {}).get('mode') == 'wall' and not product.get('door')


def area(slot, product, room):
    anchor = slot.wallMount
    length = room.width if anchor.wall in {'north', 'south'} else room.depth
    start = .05 + (length - product['width'] - .1) * anchor.offset
    return start, start + product['width'], anchor.height - product['height'] / 2, anchor.height + product['height'] / 2


def normalize(slot, product, room):
    require(slot.wallMount is not None, 'wall_mount', 'Wall objects need a wall anchor.')
    a, b, bottom, _ = area(slot, product, room)
    center, inset, wall = (a+b)/2, product['depth']/2, slot.wallMount.wall
    slot.x = (inset if wall == 'west' else room.width-inset if wall == 'east' else center) - room.width/2
    slot.z = (inset if wall == 'north' else room.depth-inset if wall == 'south' else room.depth-center) - room.depth/2
    slot.rotation = {'north': 0, 'west': 90, 'south': 180, 'east': 270}[wall]
    slot.elevation, slot.supportId = bottom, None


def validate(state, products):
    from backend.placement import opening, pose
    for slot in state.slots.values():
        product = products.get(slot.catalogId)
        if not product:
            continue
        if not is_wall_fixture(product):
            require(slot.wallMount is None, 'wall_mount', 'Only wall objects can use fixture anchors.')
            continue
        require(slot.wallMount is not None and not slot.door and not slot.supportId, 'wall_mount', 'Wall objects need their own wall anchor.')
        normal = slot.model_copy(deep=True)
        normalize(normal, product, state.room)
        require(pose(normal) == pose(slot), 'wall_mount', 'Object position must match its wall anchor.')
        wall = slot.wallMount.wall
        length = state.room.width if wall in {'north', 'south'} else state.room.depth
        a, b, bottom, top = area(slot, product, state.room)
        require(a >= .045 and b <= length-.045 and bottom >= .045 and top <= state.room.height-.045,
                'wall_bounds', 'Keep the whole object within the wall.')
        openings = [opening(state.room, w.wall, w.offset, w.width, w.height, w.sill) for w in state.room.windows if w.wall == wall]
        for other in state.slots.values():
            p = products.get(other.catalogId)
            if p and p.get('door') and other.door and other.door.wall == wall:
                openings.append(opening(state.room, wall, other.door.offset, p['width'], p['height']))
        for c, d, low, high in openings:
            require(a >= d+.05 or b <= c-.05 or bottom >= high+.05 or top <= low-.05,
                    'wall_opening', 'Keep wall objects clear of windows and doors.')

File: frontend/src/wallFixtures.ts
import type { Item, Product, Scene } from './catalog.ts'
import { doorGeometry } from './doors.ts'
import { defaultWindows, walls, windowGeometry } from './sunlight.ts'

const GAP = .05
const EPSILON = .005

export function isWallFixture(product: Product) {
  return product.placement?.mode === 'wall' && !product.door
}

export function wallFixtureGeometry(item: Item, product: Product, scene: Pick<Scene, 'width' | 'depth'>) {
  const anchor = item.wallMount!
  const length = anchor.wall === 'north' || anchor.wall === 'south' ? scene.width : scene.depth
  const [width, height] = product.dimensions
  const start = GAP + (length - width - 2 * GAP) * anchor.offset
  return { wall: anchor.wall, length, start, end: start + width,
    bottom: anchor.height - height / 2, top: anchor.height + height / 2 }
}

/** Local +Z faces into the room; the back of the fixture is flush to the wall. */
export function normalizeWallFixture(item: Item, product: Product, scene: Pick<Scene, 'width' | 'depth'>): Item {
  if (!isWallFixture(product) || !item.wallMount) return item
  const opening = wallFixtureGeometry(item, product, scene)
  const center = (opening.start + opening.end) / 2
  const inset = product.dimensions[2] / 2
  const wall = opening.wall
  return { ...item,
    x: (wall === 'west' ? inset : wall === 'east' ? scene.width - inset : center) - scene.width / 2,
    z: (wall === 'north' ? inset : wall === 'south' ? scene.depth - inset : scene.depth - center) - scene.depth / 2,
    rotation: wall === 'north' ? 0 : wall === 'west' ? 90 : wall === 'south' ? 180 : 270,
    elevation: opening.bottom, supportId: undefined,
  }
}

export function validWallFixture(item: Item, product: Product, scene: Scene, catalog: Product[]) {
  const anchor = item.wallMount
  if (!isWallFixture(product) || !anchor || !walls.includes(anchor.wall) ||
    !Number.isFinite(anchor.offset) || anchor.offset < 0 || anchor.offset > 1 ||
    !Number.isFinite(anchor.height) || item.supportId || item.door) return false
  const area = wallFixtureGeometry(item, product, scene)
  if (![scene.width, scene.depth, item.x, item.z, item.rotation, item.elevation, ...product.dimensions].every(Number.isFinite) ||
    area.start < GAP - EPSILON || area.end > area.length - GAP + EPSILON ||
    area.bottom < GAP - EPSILON || area.top > (scene.height ?? 2.6) - GAP + EPSILON) return false
  const normalized = normalizeWallFixture(item, product, scene)
  if (Math.abs(normalized.x - item.x) > EPSILON || Math.abs(normalized.z - item.z) > EPSILON ||
    Math.abs(normalized.elevation! - item.elevation!) > EPSILON || normalized.rotation !== item.rotation) return false
  const openings = [
    ...(scene.windows ?? defaultWindows).filter(window => window.wall === anchor.wall).map(window => windowGeometry(window, scene)),
    ...scene.items.flatMap(other => {
      const otherProduct = catalog.find(p => p.id === other.productId)
      return otherProduct?.door && other.door?.wall === anchor.wall ? [doorGeometry(other, otherProduct, scene)] : []
    }),
  ]
  return openings.every(other => area.start >= other.end + GAP || area.end <= other.start - GAP ||
    area.bottom >= other.top + GAP || area.top <= other.bottom - GAP)
}
